pkg/k8s/factory.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"reflect"
+	"time"
+
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/serializer"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	networkingv1 "k8s.io/api/networking/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ControllerFactory returns a kubernetes controller.
+// Parameters:
+//  * k8sGetter is the client used to watch for kubernetes events.
+//  * resourceObj: is an object of the type that you expect to receive.
+//  * resyncPeriod: if non-zero, will re-list this often (you will get OnUpdate
+//    calls, even if nothing changed). Otherwise, re-list will be delayed as
+//    long as possible (until the upstream source closes the watch or times out,
+//    or you stop the controller).
+//  * rehf: is the  resource event handler funcs that is used to handle the
+//    stream events.
+func ControllerFactory(
+	k8sGetter cache.Getter,
+	resourceObj runtime.Object,
+	resyncPeriod time.Duration,
+	rehf cache.ResourceEventHandlerFuncs,
+) cache.Controller {
+
+	_, controller := cache.NewInformer(
+		cache.NewListWatchFromClient(
+			k8sGetter,
+			resourceNameOf(resourceObj),
+			v1.NamespaceAll,
+			fields.Everything(),
+		),
+		resourceObj,
+		resyncPeriod,
+		rehf,
+	)
+
+	return controller
+}
+
+// ResourceEventHandlerFactory returns a kubernetes ResourceEventHandlerFuncs,
+// the resource event handler will have a serializer.FunctionQueue to enqueue
+// the received events.
+// Parameters:
+//  * addFunc is the function serialized in the queue when an object addition is
+//    received.
+//  * delFunc is the function serialized in the queue when an object deletion is
+//    received.
+//  * updateFunc is the function serialized in the queue when an object update
+//    is received.
+//  * resourceObj: is an object of the type that you expect to receive.
+func ResourceEventHandlerFactory(
+	addFunc, delFunc func(i interface{}) func() error,
+	updateFunc func(old, new interface{}) func() error,
+	resourceObj runtime.Object,
+) cache.ResourceEventHandlerFuncs {
+
+	fqueue := serializer.NewFunctionQueue(1024)
```

Comment:
It might be worth documenting the queue-length change in a commit message. It shouldn't matter but it will change how often Enqueue blocks.